File: CSharp/One/Ast/AstTypes.cs
using One.Ast;

namespace One.Ast
{
    public interface IPrimitiveType : IType {
        
    }
    
    public interface IHasTypeArguments {
        IType[] typeArguments { get; set; }
    }
    
    public interface IInterfaceType : IType {
        IType[] typeArguments { get; set; }
        
        IInterface getDecl();
    }
    
    public class TypeHelper {
        public static string argsRepr(IType[] args)
        {
            return args.length() == 0 ? "" : $"<{args.map(x => x.repr()).join(", ")}>";
        }
        
        public static bool isGeneric(IType type)
        {
            if (type is GenericsType)
                return true;
            else if (type is ClassType classType)
                return classType.typeArguments.some(x => TypeHelper.isGeneric(x));
            else if (type is InterfaceType intType)
                return intType.typeArguments.some(x => TypeHelper.isGeneric(x));
            else if (type is LambdaType lambdType)
                return lambdType.parameters.some(x => TypeHelper.isGeneric(x.type)) || TypeHelper.isGeneric(lambdType.returnType);
            else
                return false;
        }
        
        public static bool equals(IType type1, IType type2)
        {
            if (type1 == null || type2 == null)
                throw new Error("Type is missing!");
            if (type1 is VoidType && type2 is VoidType)
                return true;
            if (type1 is AnyType && type2 is AnyType)
                return true;
            if (type1 is GenericsType genType && type2 is GenericsType genType2)
                return genType.typeVarName == genType2.typeVarName;
            if (type1 is EnumType enumType && type2 is EnumType enumType2)
                return enumType.decl == enumType2.decl;
            if (type1 is LambdaType lambdType2 && type2 is LambdaType lambdType3)
                return TypeHelper.equals(lambdType2.returnType, lambdType3.returnType) && lambdType2.parameters.length() == lambdType3.parameters.length() && lambdType2.parameters.every((t, i) => TypeHelper.equals(t.type, lambdType3.parameters.get(i).type));
            if (type1 is ClassType classType2 && type2 is ClassType classType3)
                return classType2.decl == classType3.decl && classType2.typeArguments.length() == classType3.typeArguments.length() && classType2.typeArguments.every((t, i) => TypeHelper.equals(t, classType3.typeArguments.get(i)));
            if (type1 is InterfaceType intType2 && type2 is InterfaceType intType3)
                return intType2.decl == intType3.decl && intType2.typeArguments.length() == intType3.typeArguments.length() && intType2.typeArguments.every((t, i) => TypeHelper.equals(t, intType3.typeArguments.get(i)));
            return false;
        }
        
        public static bool isAssignableTo(IType toBeAssigned, IType whereTo)
        {
            // AnyType can assigned to any type except to void
            if (toBeAssigned is AnyType && !(whereTo is VoidType))
                return true;
            // any type can assigned to AnyType except void
            if (whereTo is AnyType && !(toBeAssigned is VoidType))
                return true;
            // any type can assigned to GenericsType except void
            if (whereTo is GenericsType && !(toBeAssigned is VoidType))
                return true;
            // null can be assigned anywhere
            // TODO: filter out number and boolean types...
            if (toBeAssigned is NullType && !(whereTo is VoidType))
                return true;
            
            if (TypeHelper.equals(toBeAssigned, whereTo))
                return true;
            
            if (toBeAssigned is ClassType classType4 && whereTo is ClassType classType5)
                return (classType4.decl.baseClass != null && TypeHelper.isAssignableTo(classType4.decl.baseClass, classType5)) || classType4.decl == classType5.decl && classType4.typeArguments.every((x, i) => TypeHelper.isAssignableTo(x, classType5.typeArguments.get(i)));
            if (toBeAssigned is ClassType classType6 && whereTo is InterfaceType intType4)
                return (classType6.decl.baseClass != null && TypeHelper.isAssignableTo(classType6.decl.baseClass, intType4)) || classType6.decl.baseInterfaces.some(x => TypeHelper.isAssignableTo(x, intType4));
            if (toBeAssigned is InterfaceType intType5 && whereTo is InterfaceType intType6)
                return intType5.decl.baseInterfaces.some(x => TypeHelper.isAssignableTo(x, intType6)) || intType5.decl == intType6.decl && intType5.typeArguments.every((x, i) => TypeHelper.isAssignableTo(x, intType6.typeArguments.get(i)));
            if (toBeAssigned is LambdaType lambdType4 && whereTo is LambdaType lambdType5)
                return lambdType4.parameters.length() == lambdType5.parameters.length() && lambdType4.parameters.every((p, i) => TypeHelper.isAssignableTo(p.type, lambdType5.parameters.get(i).type)) && (TypeHelper.isAssignableTo(lambdType4.returnType, lambdType5.returnType) || lambdType5.returnType is GenericsType);
            
            return false;
        }
    }
    
    public class VoidType : IPrimitiveType {
        public static VoidType instance;
        
        static VoidType()
        {
            VoidType.instance = new VoidType();
        }
        
        public string repr()
        {
            return "Void";
        }
    }
    
    public class AnyType : IPrimitiveType {
        public static AnyType instance;
        
        static AnyType()
        {
            AnyType.instance = new AnyType();
        }
        
        public string repr()
        {
            return "Any";
        }
    }
    
    public class NullType : IPrimitiveType {
        public static NullType instance;
        
        static NullType()
        {
            NullType.instance = new NullType();
        }
        
        public string repr()
        {
            return "Null";
        }
    }
    
    public class GenericsType : IType {
        public string typeVarName;
        
        public GenericsType(string typeVarName)
        {
            this.typeVarName = typeVarName;
        }
        
        public string repr()
        {
            return $"G:{this.typeVarName}";
        }
    }
    
    public class EnumType : IType {
        public Enum_ decl;
        
        public EnumType(Enum_ decl)
        {
            this.decl = decl;
        }
        
        public string repr()
        {
            return $"E:{this.decl.name}";
        }
    }
    
    public class InterfaceType : IType, IHasTypeArguments, IInterfaceType {
        public Interface decl;
        public IType[] typeArguments { get; set; }
        
        public InterfaceType(Interface decl, IType[] typeArguments)
        {
            this.decl = decl;
            this.typeArguments = typeArguments;
        }
        
        public IInterface getDecl()
        {
            return this.decl;
        }
        
        public string repr()
        {
            return $"I:{this.decl.name}{TypeHelper.argsRepr(this.typeArguments)}";
        }
    }
    
    public class ClassType : IType, IHasTypeArguments, IInterfaceType {
        public Class decl;
        public IType[] typeArguments { get; set; }
        
        public ClassType(Class decl, IType[] typeArguments)
        {
            this.decl = decl;
            this.typeArguments = typeArguments;
        }
        
        public IInterface getDecl()
        {
            return this.decl;
        }
        
        public string repr()
        {
            return $"C:{this.decl.name}{TypeHelper.argsRepr(this.typeArguments)}";
        }
    }
    
    public class UnresolvedType : IType, IHasTypeArguments {
        public string typeName;
        public IType[] typeArguments { get; set; }
        
        public UnresolvedType(string typeName, IType[] typeArguments)
        {
            this.typeName = typeName;
            this.typeArguments = typeArguments;
        }
        
        public string repr()
        {
            return $"X:{this.typeName}{TypeHelper.argsRepr(this.typeArguments)}";
        }
    }
    
    public class LambdaType : IType {
        public MethodParameter[] parameters;
        public IType returnType;
        
        public LambdaType(MethodParameter[] parameters, IType returnType)
        {
            this.parameters = parameters;
            this.returnType = returnType;
        }
        
        public string repr()
        {
            return $"L:({this.parameters.map(x => x.type.repr()).join(", ")})=>{this.returnType.repr()}";
        }
    }
}

File: PHP/One/Ast/AstTypes.php
<?php

namespace One\Ast\AstTypes;

use One\Ast\Types\Enum;
use One\Ast\Types\Interface_;
use One\Ast\Types\Class_;
use One\Ast\Types\MethodParameter;
use One\Ast\Types\IInterface;
use One\Ast\Interfaces\IType;

interface IPrimitiveType extends IType {
    
}

interface IHasTypeArguments {
    
}

interface IInterfaceType extends IType {
    function getDecl();
}

class TypeHelper {
    static function argsRepr($args) {
        return count($args) === 0 ? "" : "<" . implode(", ", array_map(function ($x) { return $x->repr(); }, $args)) . ">";
    }
    
    static function isGeneric($type) {
        if ($type instanceof GenericsType)
            return true;
        else if ($type instanceof ClassType)
            return \OneLang\ArrayHelper::some($type->typeArguments, function ($x) { return TypeHelper::isGeneric($x); });
        else if ($type instanceof InterfaceType)
            return \OneLang\ArrayHelper::some($type->typeArguments, function ($x) { return TypeHelper::isGeneric($x); });
        else if ($type instanceof LambdaType)
            return \OneLang\ArrayHelper::some($type->parameters, function ($x) { return TypeHelper::isGeneric($x->type); }) || TypeHelper::isGeneric($type->returnType);
        else
            return false;
    }
    
    static function equals($type1, $type2) {
        if ($type1 === null || $type2 === null)
            throw new \OneLang\Error("Type is missing!");
        if ($type1 instanceof VoidType && $type2 instanceof VoidType)
            return true;
        if ($type1 instanceof AnyType && $type2 instanceof AnyType)
            return true;
        if ($type1 instanceof GenericsType && $type2 instanceof GenericsType)
            return $type1->typeVarName === $type2->typeVarName;
        if ($type1 instanceof EnumType && $type2 instanceof EnumType)
            return $type1->decl === $type2->decl;
        if ($type1 instanceof LambdaType && $type2 instanceof LambdaType)
            return TypeHelper::equals($type1->returnType, $type2->returnType) && count($type1->parameters) === count($type2->parameters) && \OneLang\ArrayHelper::every($type1->parameters, function ($t, $i) use ($type2) { return TypeHelper::equals($t->type, $type2->parameters[$i]->type); });
        if ($type1 instanceof ClassType && $type2 instanceof ClassType)
            return $type1->decl === $type2->decl && count($type1->typeArguments) === count($type2->typeArguments) && \OneLang\ArrayHelper::every($type1->typeArguments, function ($t, $i) use ($type2) { return TypeHelper::equals($t, $type2->typeArguments[$i]); });
        if ($type1 instanceof InterfaceType && $type2 instanceof InterfaceType)
            return $type1->decl === $type2->decl && count($type1->typeArguments) === count($type2->typeArguments) && \OneLang\ArrayHelper::every($type1->typeArguments, function ($t, $i) use ($type2) { return TypeHelper::equals($t, $type2->typeArguments[$i]); });
        return false;
    }
    
    static function isAssignableTo($toBeAssigned, $whereTo) {
        // AnyType can assigned to any type except to void
        if ($toBeAssigned instanceof AnyType && !($whereTo instanceof VoidType))
            return true;
        // any type can assigned to AnyType except void
        if ($whereTo instanceof AnyType && !($toBeAssigned instanceof VoidType))
            return true;
        // any type can assigned to GenericsType except void
        if ($whereTo instanceof GenericsType && !($toBeAssigned instanceof VoidType))
            return true;
        // null can be assigned anywhere
        // TODO: filter out number and boolean types...
        if ($toBeAssigned instanceof NullType && !($whereTo instanceof VoidType))
            return true;
        
        if (TypeHelper::equals($toBeAssigned, $whereTo))
            return true;
        
        if ($toBeAssigned instanceof ClassType && $whereTo instanceof ClassType)
            return ($toBeAssigned->decl->baseClass !== null && TypeHelper::isAssignableTo($toBeAssigned->decl->baseClass, $whereTo)) || $toBeAssigned->decl === $whereTo->decl && \OneLang\ArrayHelper::every($toBeAssigned->typeArguments, function ($x, $i) use ($whereTo) { return TypeHelper::isAssignableTo($x, $whereTo->typeArguments[$i]); });
        if ($toBeAssigned instanceof ClassType && $whereTo instanceof InterfaceType)
            return ($toBeAssigned->decl->baseClass !== null && TypeHelper::isAssignableTo($toBeAssigned->decl->baseClass, $whereTo)) || \OneLang\ArrayHelper::some($toBeAssigned->decl->baseInterfaces, function ($x) use ($whereTo) { return TypeHelper::isAssignableTo($x, $whereTo); });
        if ($toBeAssigned instanceof InterfaceType && $whereTo instanceof InterfaceType)
            return \OneLang\ArrayHelper::some($toBeAssigned->decl->baseInterfaces, function ($x) use ($whereTo) { return TypeHelper::isAssignableTo($x, $whereTo); }) || $toBeAssigned->decl === $whereTo->decl && \OneLang\ArrayHelper::every($toBeAssigned->typeArguments, function ($x, $i) use ($whereTo) { return TypeHelper::isAssignableTo($x, $whereTo->typeArguments[$i]); });
        if ($toBeAssigned instanceof LambdaType && $whereTo instanceof LambdaType)
            return count($toBeAssigned->parameters) === count($whereTo->parameters) && \OneLang\ArrayHelper::every($toBeAssigned->parameters, function ($p, $i) use ($whereTo) { return TypeHelper::isAssignableTo($p->type, $whereTo->parameters[$i]->type); }) && (TypeHelper::isAssignableTo($toBeAssigned->returnType, $whereTo->returnType) || $whereTo->returnType instanceof GenericsType);
        
        return false;
    }
}

class VoidType implements IPrimitiveType {
    public static $instance;
    
    static function StaticInit()
    {
        VoidType::$instance = new VoidType();
    }
    
    function repr() {
        return "Void";
    }
}
VoidType::StaticInit();

class AnyType implements IPrimitiveType {
    public static $instance;
    
    static function StaticInit()
    {
        AnyType::$instance = new AnyType();
    }
    
    function repr() {
        return "Any";
    }
}
AnyType::StaticInit();

class NullType implements IPrimitiveType {
    public static $instance;
    
    static function StaticInit()
    {
        NullType::$instance = new NullType();
    }
    
    function repr() {
        return "Null";
    }
}
NullType::StaticInit();

class GenericsType implements IType {
    public $typeVarName;
    
    function __construct($typeVarName) {
        $this->typeVarName = $typeVarName;
    }
    
    function repr() {
        return "G:" . $this->typeVarName;
    }
}

class EnumType implements IType {
    public $decl;
    
    function __construct($decl) {
        $this->decl = $decl;
    }
    
    function repr() {
        return "E:" . $this->decl->name;
    }
}

class InterfaceType implements IType, IHasTypeArguments, IInterfaceType {
    public $decl;
    public $typeArguments;
    
    function __construct($decl, $typeArguments) {
        $this->decl = $decl;
        $this->typeArguments = $typeArguments;
    }
    
    function getDecl() {
        return $this->decl;
    }
    
    function repr() {
        return "I:" . $this->decl->name . TypeHelper::argsRepr($this->typeArguments);
    }
}

class ClassType implements IType, IHasTypeArguments, IInterfaceType {
    public $decl;
    public $typeArguments;
    
    function __construct($decl, $typeArguments) {
        $this->decl = $decl;
        $this->typeArguments = $typeArguments;
    }
    
    function getDecl() {
        return $this->decl;
    }
    
    function repr() {
        return "C:" . $this->decl->name . TypeHelper::argsRepr($this->typeArguments);
    }
}

class UnresolvedType implements IType, IHasTypeArguments {
    public $typeName;
    public $typeArguments;
    
    function __construct($typeName, $typeArguments) {
        $this->typeName = $typeName;
        $this->typeArguments = $typeArguments;
    }
    
    function repr() {
        return "X:" . $this->typeName . TypeHelper::argsRepr($this->typeArguments);
    }
}

class LambdaType implements IType {
    public $parameters;
    public $returnType;
    
    function __construct($parameters, $returnType) {
        $this->parameters = $parameters;
        $this->returnType = $returnType;
    }
    
    function repr() {
        return "L:(" . implode(", ", array_map(function ($x) { return $x->type->repr(); }, $this->parameters)) . ")=>" . $this->returnType->repr();
    }
}
